File: jupyterlab/devel/extractSchema.js
const path = require('path')
const fs = require('fs')


function main() {
  const schema = require('../jsonschema/root.json')
  dumpOne(schema, ['MessageToStellarGlobe'], './jsonschema/MessageToStellraGlobe.json')
  dumpChildren(schema, ['MessageToJupyterLabMap'], './jsonschema/MessageToJupyterLab')
  dumpChildren(schema, ['MessageToStellarGlobeMap'], './jsonschema/MessageToStellraGlobe')
  dumpChildren(schema, ['LayerPropsMap'], './jsonschema/LayerProps')
}


function dumpOne(schema, routes, out) {
  fs.mkdirSync(path.dirname(out), { recursive: true })
  fs.writeFileSync(
    out,
    JSON.stringify(extractSchema(schema, routes), null, 2),
  )
}


function dumpChildren(schema, routes, directory) {
  const parent = extractSchema(schema, routes)
  fs.mkdirSync(directory, { recursive: true })
  Object.keys(parent.properties).forEach(child => {
    fs.writeFileSync(
      `${directory}/${child}.json`,
      JSON.stringify(extractSchema(parent, [child]), null, 2),
    )
  })
}


function extractSchema(schema, routes) {
  function dig(schema, routes, definitions = {}) {
    definitions = { ...schema.definitions ?? {}, ...definitions }
    schema = dereference(schema, definitions)
    if (routes.length === 0) {
      return schema
    }
    else {
      const [route, ...rest] = routes
      const nextSchema = schema.properties[route]
      return dig(nextSchema, rest, definitions)
    }
  }

  function isRef(obj) {
    return obj && !!obj.$ref
  }

  function dereference(obj, definitions) {
    if (isRef(obj)) {
      /** @type {string} */
      const $ref = obj.$ref
      console.assert($ref.startsWith('#/definitions/'))
      return definitions[definitionKey($ref)]
    }
    return obj
  }

  function definitionKey($ref) {
    return $ref.slice('#/definitions/'.length)
  }


  function cleanupDefinitions(schema, definitions) {
    const usedDefinitaions = new Set()
    const walk = (o) => {
      if (isRef(o)) {
        const key = definitionKey(o.$ref)
        if (!usedDefinitaions.has(key)) {
          usedDefinitaions.add(definitionKey(o.$ref))
          walk(definitions[key], definitions)
        }
      }
      else if (o instanceof Array) {
        o.forEach(c => walk(c))
      }
      else if (o instanceof Object) {
        Object.keys(o).forEach(k => {
          if (k !== 'definitions') {
            walk(o[k])
          }
        })
      }
    }
    walk(schema)
    return Object.fromEntries(
      Array.from(usedDefinitaions.values()).map(k => [k, definitions[k]])
    )
  }

  const target = dig(schema, routes)
  const newSchema = {
    ...target,
    definitions: cleanupDefinitions(target, schema.definitions),
  }
  return newSchema
}


main()

File: python-integration/python/devel/extractschema.py
from typing import List


def extractSchema(schema, routes: List[str]):
    # machine translation of devel/generate-js-type-validators.js

    def dig(schema, routes, definitions=None):
        if definitions is None:
            definitions = {}
        definitions = {**schema.get('definitions', {}), **definitions}
        schema = dereference(schema, definitions)

        if len(routes) == 0:
            return schema
        else:
            route, *rest = routes
            nextSchema = schema['properties'][route]
            return dig(nextSchema, rest, definitions)

    def isRef(obj):
        return isinstance(obj, dict) and '$ref' in obj

    def dereference(obj, definitions):
        if isRef(obj):
            ref = obj['$ref']
            assert ref.startswith('#/definitions/')
            return definitions[definitionKey(ref)]
        return obj

    def definitionKey(ref):
        return ref[len('#/definitions/') :]

    def cleanupDefinitions(schema, definitions):
        used_definitions = set()

        def walk(o):
            nonlocal used_definitions
            if isRef(o):
                key = definitionKey(o['$ref'])
                if key not in used_definitions:
                    used_definitions.add(key)
                    walk(definitions[key])
            elif isinstance(o, list):
                for c in o:
                    walk(c)
            elif isinstance(o, dict):
                for k, v in o.items():
                    if k != 'definitions':
                        walk(v)

        walk(schema)
        return {k: definitions[k] for k in used_definitions}

    target = dig(schema, routes)
    newSchema = {**target, 'definitions': cleanupDefinitions(target, schema.get('definitions', {}))}
    return newSchema
